### DISCUSSION TITLE
Does tokio have a native way to do MPMC (multiple producers / multiple consumers)?

### DISCUSSION THREAD

**Discussion by ilmoi** [tokio-rs/tokio#3891] [Q&A] [answered]
The best I've been able to find is [broadcast](https://docs.rs/tokio/0.2.6/tokio/sync/broadcast/index.html) - but that sends the message to all consumers and waits for all of them to receive it.

What I'm after is more of a queue based system where numerous producers stuff the pipeline with tasks and numerous consumers pick the tasks off, one by one, and complete them.

Does tokio have such functionality? If not, what are some other recommended crates?

**Comment by Darksonn** [comment, marked as answer]
Tokio does not provide a traditional mpmc channel where each message is seen by exactly one receiver. If you need this, I recommend that you use the [`async-channel`][2] or [`flume`][1] crates.

[1]: https://docs.rs/flume/
[2]: https://docs.rs/async-channel/

**Reply by ilmoi** [reply]
awesome, thanks

**Comment by stevenleadbeater** [comment]


**Reply by Darksonn** [reply]
The broadcast channel sends the message to all receivers, which is not what the question asker wants. They want each message to go only to one receiver.

**Reply by stevenleadbeater** [reply]
deleting the comment to save confusion, sorry!
